drivers/ipm/ipm_adsp.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) 2019, Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <device.h>
+#include <soc.h>
+#include <drivers/ipm.h>
+
+#include <platform/mailbox.h>
+#include <platform/shim.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ipm_adsp, CONFIG_IPM_LOG_LEVEL);
+
+/* FIXME define in DTS */
+#define DT_IPM_ADSP_0_BASE_ADDRESS	0x10
+#define DT_IPM_ADSP_0_NAME		"IPM_0"
+#define DT_IPM_ADSP_0_IRQ		SOC_AGGREGATE_IRQ(7, 6)
+#define DT_IPM_ADSP_0_IRQ_PRI		3
+
+/* FIXME: set to 64 for now */
+#define IPM_ADSP_MAX_DATA_SIZE		256
+#define IPM_ADSP_MAX_ID_VAL		IPC_DIPCI_MSG_MASK
+
+/* Mailbox ADSP -> Host */
+#define IPM_ADSP_MAILBOX_OUT		MAILBOX_DSPBOX_BASE
+#define IPM_ADSP_MAILBOX_OUT_SIZE	MAILBOX_DSPBOX_SIZE
+
+/* Mailbox Host -> ADSP */
+#define IPM_ADSP_MAILBOX_IN		MAILBOX_HOSTBOX_BASE
+#define IPM_ADSP_MAILBOX_IN_SIZE	MAILBOX_HOSTBOX_SIZE
+
+struct ipm_adsp_config {
+	u8_t *base;
+	void (*irq_config_func)(struct device *dev);
+};
+
+struct ipm_adsp_data {
+	ipm_callback_t callback;
+	void *callback_ctx;
+};
+
+static void ipm_adsp_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct ipm_adsp_data *data = dev->driver_data;
+	uint32_t dipcctl, dipcie, dipct;
+
+	dipct = ipc_read(IPC_DIPCT);
+	dipcie = ipc_read(IPC_DIPCIE);
+	dipcctl = ipc_read(IPC_DIPCCTL);
+
+	LOG_DBG("dipct 0x%x dipcie 0x%x dipcctl 0x%x", dipct, dipcie, dipcctl);
+
+	/* IPC Target Busy Interrupt Enable (IPCTBIE) */
+	if (dipct & IPC_DIPCT_BUSY && dipcctl & IPC_DIPCCTL_IPCTBIE) {
+		/* Mask Busy interrupt */
+		ipc_write(IPC_DIPCCTL, dipcctl & ~IPC_DIPCCTL_IPCTBIE);
```

Comment:
So... you're in the ISR here already.  Why is this masking the interrupt it's servicing?  I'm guessing that this is either needless, or what is happening here is you're masking delivery of the interrupt to the OTHER side of the IPM channel while the shared state is indeterminate, in which case you really need a comment here (and probably need some other locking I'm guessing?).  Or maybe the mask/unmask sequence is needed to properly ACK the interrupt?  Same thing, needs a comment.

Reply:
I do not have good documentation for this, only register description. I have followed what SOF is doing for IPC IRQ handler:
https://github.com/thesofproject/sof/blob/3dd843a46defd5234a517991a7c0049c66233b1d/src/drivers/intel/cavs/ipc.c#L64-L137
I will add more comments.

Reply:
If I cast my mind back, we cant clear this IRQ until we have finished processing the IPC (which may take some time) hence the mask.  Clearing could also interrupt the host on some HW too. So short term this flow is fine as it survives 25k flood IPC pings. Long term we should be operating a msg queue where we can clear the IRQ and process the IPC offline (and allow other IPC to be sent).